src/midgard/pointll.cc

```diff
@@ -124,77 +124,73 @@ float PointLL::Heading(const PointLL& ll2) const {
-
-    // Project a onto b where b is the origin vector representing this segment
-    // and a is the origin vector to the point we are projecting, (a.b/b.b)*b
-    auto bx = v.lng() - u.lng();
-    auto by = v.lat() - u.lat();
-
-    // Scale longitude when finding the projection. Only need the numerator
-    // at first, saving the division for the case where the lat,lon projects
-    // along the current segment.
-    auto bx2 = bx * lon_scale;
-    auto sq = bx2 * bx2 + by * by;
-    auto scale = (lng() - u.lng()) * lon_scale * bx2 + (lat() - u.lat()) * by;
-
-    // Projects along the ray before u
-    if (scale <= 0.f) {
-      point = {u.lng(), u.lat()};
-    } // Projects along the ray after v
-    else if (scale >= sq) {
-      point = {v.lng(), v.lat()};
-    } // Projects along the ray between u and v
-    else {
-      scale /= sq;
-      point = {u.lng() + bx * scale, u.lat() + by * scale};
-    }
+  float mindistsqr = approx.DistanceSquared(closest);
+
+  // start going backwards, then go forwards
+  for (int reverse = 0; reverse < 2; ++reverse) {
+    // get the range and distance for this direction
+    auto dist_cutoff = reverse ? reverse_dist_cutoff : forward_dist_cutoff;
+    int increment = reverse ? -1 : 1;
+    int indices = reverse ? pivot_index : (pts.size() - 1) - pivot_index;
+
+    PointLL point;
+    for (int index = pivot_index; indices > 0 && dist_cutoff > 0.f; index += increment, --indices) {
+      // Get the current segment
+      const PointLL& u = pts[index];
+      const PointLL& v = pts[index + increment];
+
+      // Project a onto b where b is the origin vector representing this segment
+      // and a is the origin vector to the point we are projecting, (a.b/b.b)*b
+      auto bx = v.lng() - u.lng();
+      auto by = v.lat() - u.lat();
+
+      // Scale longitude when finding the projection. Avoid divided-by-zero
+      // which gives a NaN scale, otherwise comparisons below will fail
+      auto bx2 = bx * approx.GetLngScale();
+      auto sq = bx2 * bx2 + by * by;
+      auto scale =
+          sq > 0 ? (((lng() - u.lng()) * approx.GetLngScale() * bx2 + (lat() - u.lat()) * by) / sq)
+                 : 0.f;
+
+      // Projects along the ray before u
+      if (scale <= 0.f) {
+        point = {u.lng(), u.lat()};
+      } // Projects along the ray after v
+      else if (scale >= 1.f) {
+        point = {v.lng(), v.lat()};
+      } // Projects along the ray between u and v
+      else {
+        point = {u.lng() + bx * scale, u.lat() + by * scale};
+      }
 
-    // Check if this point is better
-    const auto sq_distance = approx.DistanceSquared(point);
-    if (sq_distance < mindistsqr) {
-      closest_segment = static_cast<int>(index);
-      mindistsqr = sq_distance;
-      closest = std::move(point);
-    }
+      // Check if this point is better
+      const auto sq_distance = approx.DistanceSquared(point);
+      if (sq_distance <= mindistsqr) {
+        closest_segment = index - reverse;
```

Comment:
`- reverse` is so that you get the left most (floor index) for the segment you are on even when searching in reverse